Upgrade to rails 7.1.1

tomhughes opened the pull request:
This upgrades to rails 7.1.0 with the huge bonus of getting rid of the `composite_primary_key` dependency by switching to the new builtin rails support for composite keys.

A test site is available at https://rails71.apis.dev.openstreetmap.org/.

gravitystorm commented on app/mailers/user_mailer.rb:
```diff
@@ -124,7 +124,7 @@ def friendship_notification(friendship)
     with_recipient_locale recipient do
       @noteurl = note_url(comment.note)
-      @place = Nominatim.describe_location(comment.note.lat, comment.note.lon, 14, I18n.locale)
+      @place = Nominatim.describe_location(comment.note.lat.to_f, comment.note.lon.to_f, 14, I18n.locale)
```
This change defeats the purpose of using `DelegateClass`, because e.g. note.lat is supposed to be indistinguishable from it being a float, apart from how we format it as a string. So I'd really like to avoid having to call `.to_f` on something that's supposed to act like a float. And I also think this might break the formatting too, since to_s will be called on the float, not on the `Coord`

I spent several hours figuring out what is going on, and narrowed it down to https://github.com/rails/rails/pull/49466 and there is a fix already at https://github.com/rails/rails/pull/49576 which was merged earlier today. We should either wait for that fix to be released, or revert this commit when it happens.

gravitystorm commented on config/environments/development.rb:
```diff
@@ -56,10 +56,8 @@
-  # Debug mode disables concatenation and preprocessing of assets.
-  # This option may cause significant delays in view rendering with a large
-  # number of complex assets.
-  config.assets.debug = true
```
This change breaks my local development environment. Without the setting, I get html generated that refers to incorrect stylesheet paths. But if I add this setting back in, everything works.

Is there a reason for it being removed? If not, I would keep it for now.

I don't know why removing the setting breaks my environment, but until I figure it out I'm concerned that it might break things for other developers too.

tomhughes replied:
It was just because it was the new default from upstream. It didn't break it for me though I agree that it may make debugging a lot harder so I was surprised upstream had removed it. I guess it's related to the change in preferred methods for asset handling so I'll revert that for us.

gravitystorm reviewed (commented):
Looks good! I've added a few comments on specific things. 

Two of which are bugs that are already fixed in the `7_1_stable` branch but I don't have any visibility into when there might be a 7.1.1 release.

tomhughes commented on test/controllers/api/nodes_controller_test.rb:
```diff
@@ -138,7 +138,7 @@ def test_create_invalid_xml
       xml = "<osm><node lat='#{lat}' lon='#{lon}' changeset='#{changeset.id}'><tag k='foo' v='#{'x' * 256}'/></node></osm>"
       put node_create_path, :params => xml, :headers => auth_header
       assert_response :bad_request, "node upload did not return bad_request status"
-      assert_equal ["NodeTag ", " v: is too long (maximum is 255 characters) (\"#{'x' * 256}\")"], @response.body.split(/[0-9]+,foo:/)
```
It's CPK related - the new rails support manages to include the node ID and version in the message while with the old CPK module it failed to add the ID and you just got a space in the message. As the ID can vary it was easy just to match the message which is what we care about.

gravitystorm replied:
But if the output from the API changes, will that break client apps that are parsing the previous output?

tomhughes replied:
I remembered the details slightly wrong... This is what it used to look like:

```
NodeTag 14,foo: v: is too long (maximum is 255 characters)
```

and now:

```
NodeTag [38, "foo"]: v: is too long (maximum is 255 characters)
```

which is all down to how rails default error handling (augmented by CPK in the old code) represents a composite key in a validation failure which I'm not sure we can do much about?

Are users really supposed to be parsing this?

mmd-osm replied:
FWIW: I invented my own error messages in cgimap for keys/values exceeding 255 chars, and nobody seemed to have an issue with it. That's for the diff upload, though.

Error messages are: _Key has more than 255 unicode characters_ and _Value has more than 255 unicode characters_.

gravitystorm replied:
I know that some clients do parse the free-text error responses (see e.g. https://github.com/openstreetmap/openstreetmap-website/issues/2227 inter alia). 

But perhaps for this particular error it's very unlikely that clients would be running into it (other than during development) since it's very likely that adding the length limit would happen client-side. It's a different situation for other messages, like when an editing conflict happens (e.g. version mismatch) which you can't handle 100% client side.

So I think it's worth us being aware that this is technically a breaking change, but we don't expect (m?)any people to notice, and if there is a problem we'll need to build some method to fully control the errors in future (i.e. not rely on rails generated error messages since these can change between versions).

tomhughes commented on config/environments/test.rb:
```diff
@@ -8,13 +8,14 @@
-  # Eager loading loads your whole application. When running a single test locally,
-  # this probably isn't necessary. It's a good idea to do in a continuous integration
-  # system, or in some way before deploying your code.
-  config.eager_load = ENV["CI"].present?
```
Looks like https://github.com/rails/rails/pull/49470 was the one that got merged. I'll update the comment.

tomhughes replied:
You're right... It looks like it got to both main and 7.1 without any PR for it.

tomhughes commented on config/environments/production.rb:
```diff
@@ -16,16 +16,14 @@
 
-  # Disable serving static files from the `/public` folder by default since
-  # Apache or NGINX already handles this.
-  config.public_file_server.enabled = ENV["RAILS_SERVE_STATIC_FILES"].present?
```
This was also just a change in upstream defaults I think - as far as I know our old code was just the default and we never use that environment variable.

You're right that we use apache to serve the files, but that means rails never sees the request so whether it is also prepared to serve them won't matter and I guess other users of the code might be running in production with different setups?

gravitystorm replied:
OK, let's stick with the new default and we can always change if necessary.

tomhughes commented on config/environments/production.rb:
```diff
@@ -16,16 +16,14 @@
+  config.public_file_server.enabled = true
 
-  # Compress JavaScripts and CSS.
-  config.assets.js_compressor = Terser.new
```
Also an upstream change but I think not a new one and I missed reverting it. Will fix.

mmd-osm commented:
> A test site is available at https://rails71.apis.dev.openstreetmap.org/.

Are there any topics we should pay attention to during testing, e.g. performance regressions, or data consistency issues due to switching to the built in composite keys, maybe?

tomhughes commented:
Not really, there's no real functional difference to the indexes or SQL that's used.

tomhughes commented:
so 7.1.1 is out now but sadly https://github.com/jumph4x/canonical-rails/pull/82 stops us upgrading to it :-(

gravitystorm commented:
> So 7.1.1 is out now but sadly [jumph4x/canonical-rails#82](https://github.com/jumph4x/canonical-rails/pull/82) stops us upgrading to it :-(

I've dived deep into this today, and I'm confident that it's just an administrative holdup with that gem. However, I'm not sure how quickly they'll be able to make a release, since although it's just one small change to their gemfile, their lack of CI is a bit of a holdup (and when they get CI working, they'll find some other issues with their test suite, all of which I've reported).

I don't think we should wait for them, so I'd suggest having our own fork of the gem with the constraint relaxed (as per that issue, but under our control just in case) and then moving on with the 7.1 upgrade here before (hopefully soon) reverting back to upstream.

tomhughes commented:
I was thinking much the same thing, that we should just fork it for now and relax the constraint, so I have now done that and rebased this so it now uses 7.1.1 and drops the two commits that are now fixed upstream.

tomhughes commented on Gemfile:
```diff
@@ -48,7 +48,6 @@ gem "active_record_union"
 gem "bootstrap", :github => "gravitystorm/bootstrap-rubygem", :branch => "dartsass_5_2_3"
 gem "bootstrap_form", "~> 5.0"
 gem "cancancan"
-# gem "composite_primary_keys", "~> 14.0.0"
```
It does actually get removed in a later commit but I can do it there instead.

gravitystorm replied:
Actually when I review the sequence of commits more carefully (sorry), I'm happy as-is.

gravitystorm commented:
Otherwise LGTM!
